File: src/storeFx/setReplayGainOptions.ts
import * as Fx from '../fx/fx';
import type { ReplayGainOptions } from '../store/state';
import { REPLAY_GAIN_LOCAL_STORAGE_KEY } from '../store/store';
import type { Store } from '../store/types';

export function setReplayGainOptions(
  updateReplayGainOptions: (prev: ReplayGainOptions) => ReplayGainOptions,
) {
  return Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) => {
    const { player } = store.getState();

    const { preAmp, preferredGain } = updateReplayGainOptions(
      player.replayGainOptions,
    );

    const newReplayGainOptions = {
      preAmp: Math.max(-15, Math.min(preAmp, 15)),
      preferredGain,
    };

    store.setState(prevState => ({
      player: {
        ...prevState.player,
        replayGainOptions: newReplayGainOptions,
      },
    }));

    localStorage.setItem(
      REPLAY_GAIN_LOCAL_STORAGE_KEY,
      JSON.stringify(newReplayGainOptions),
    );

    return Fx.Ok();
  });
}

File: src/storeFx/setReplayGainOptions.ml
open Dom.Storage
open Fx
open Monad_syntax
open Store.State

let make newReplayGainOptions =
  let* Deps.{ store } = ask () in
  let state = store |. Zustand.getState in
  let newReplayGainOptions =
    state.player.replayGainOptions |> newReplayGainOptions
  in
  let newReplayGainOptions =
    ReplayGainOptions.
      {
        newReplayGainOptions with
        preAmp =
          Js.Math.max_float (-15.0)
            (Js.Math.min_float newReplayGainOptions.preAmp 15.0);
      }
  in
  store
  |. Zustand.setState (fun prevState ->
      {
        prevState with
        player =
          { prevState.player with replayGainOptions = newReplayGainOptions };
      });
  Ok
    (newReplayGainOptions |> Js.Json.stringifyAny
    |> Option.iter (fun value ->
        localStorage |> setItem Store.Player.replay_gain_local_storage_key value)
    )
